lib/models/per-bundle-addon-cache/index.js

```diff
@@ -0,0 +1,402 @@
+          this.numProxies++;
+          return getAddonProxy(addonCacheEntry, parent);
+        } else {
+          // the cache entry was created 'empty' by an earlier call, indicating
+          // an addon that is used in a lazy engine but also used in the project,
+          // and we're now creating the instance for the project.
+          // Fill in the entry and return the new instance.
+          logger.debug(`About to fill in REGULAR ADDON EXISTING cache entry for addon at: ${addonPkgInfo.realPath}`);
+          this.numAddonInstances++;
+          addonInstance = addonPkgInfo.constructAddonInstance(parent, this.project);
+          addonCacheEntry[TARGET_INSTANCE] = addonInstance; // cache BEFORE initializing child addons
+          addonPkgInfo.initChildAddons(addonInstance);
+          return addonInstance;
+        }
+      }
+
+      // There is no entry for this addon in the bundleHost's addon cache. Create a new
+      // instance, cache it in the addon cache, and return it.
+      logger.debug(`About to construct REGULAR ADDON NEW cache entry for addon at: ${addonPkgInfo.realPath}`);
+      this.numAddonInstances++;
+      addonInstance = addonPkgInfo.constructAddonInstance(parent, this.project);
+      addonCacheEntry = this.createAddonCacheEntry(addonInstance, addonPkgInfo.realPath);
+      addonInstanceCache.set(addonPkgInfo.realPath, addonCacheEntry); // cache BEFORE initializing child addons
+      addonPkgInfo.initChildAddons(addonInstance);
+      return addonInstance;
+    } else {
+      // The bundleHost is not the project but the project bundles the addon too and
+      // deduping is enabled, so the cache entry needs to go in the project's cache.
+      // Get/create an empty cache entry and return a proxy to it. The project will
+      // set the instance later (see above).
+      let bundleHostCacheEntry = this.bundleHostCache.get(PROJECT_BUNDLE_HOST_NAME);
+      let addonCacheEntry = bundleHostCacheEntry.addonInstanceCache.get(addonPkgInfo.realPath);
+      if (!addonCacheEntry) {
+        logger.debug(`About to construct REGULAR ADDON EMPTY cache entry for addon at: ${addonPkgInfo.realPath}`);
+        addonCacheEntry = this.createAddonCacheEntry(null, addonPkgInfo.realPath);
+        bundleHostCacheEntry.addonInstanceCache.set(addonPkgInfo.realPath, addonCacheEntry);
+      }
+
+      logger.debug(`About to construct REGULAR ADDON PROXY for EMPTY addon at: ${addonPkgInfo.realPath}`);
+      this.numProxies++;
+      return getAddonProxy(addonCacheEntry, parent);
+    }
+  }
+
+  /**
+   * Called by engine-addon when it needs to retrieve the real child addons for a given
+   * object (i.e., during ancestorHostAddons), even though the object might be a proxy.
+   *
+   * @method getAddonAddons
+   * @param {Addon|Proxy} addonInstance an instance of an addon or a Proxy generated by this class
+   * @return {Array} an array of addon instances
+   */
+  getAddonAddons(addonInstance) {
```

Comment:
why does this method live on `PerBundleAddonCache`? i'd probably just remove this for now, especially w/ plans to move over host addons/LCA host logic to `ember-cli`: https://github.com/ember-cli/ember-cli/issues/9501

Reply:
It's because I didn't want to have to completely replace the ancestorHostAddons function in ember-engines. This is what I was talking about when we discussed releasing ember-engines first (so that its function can call back into here.) The code there just says:

if (project.perBundleAddonCache) {
  addons = perBundleAddonCache.getAddonAddons(addon);
} else {
  addons = addon.addons;
}

Reply:
Until we have the fix to move the `findLCAHost` and other stuff in `ancestorHostAddons` over from ember-engines to ember-cli, we need some sort of call like the above. I'll hold this open until we decide what to do and how to stage the changes.